tensorboard/plugins/scalar/polymer3/tf_scalar_dashboard/tf-scalar-card.ts

```diff
@@ -257,19 +261,56 @@ export class TfScalarCard extends PolymerElement {
   // this.requestManager.request(
   //      this.getDataLoadUrl({tag, run, experiment})
   @property({type: Object})
-  requestData: RequestDataCallback<RunTagItem, ScalarDatum[]> = (
+  requestData: RequestDataCallback<RunTagItem, ScalarDatum[] | null> = (
     items,
     onLoad,
     onFinish
   ) => {
     const router = getRouter();
-    const baseUrl = router.pluginRoute('scalars', '/scalars');
+    const url = router.pluginRoute('scalars', '/scalars_multirun');
+    const runsByTag = new Map<string, string[]>();
+    for (const {tag, run} of items) {
+      let runs = runsByTag.get(tag);
+      if (runs == null) {
+        runsByTag.set(tag, (runs = []));
+      }
+      runs.push(run);
+    }
+
+    // Request at most this many runs at once.
+    //
+    // Back-of-the-envelope math: each scalar datum JSON value contains
+    // two floats and a small-ish integer. Floats are about 18 bytes,
+    // since f64s have -log_10(2^-53) ~= 16 digits of precision plus
+    // decimal point and leading zero. Small-ish integers (steps) are
+    // about 5 bytes. Add JSON overhead `[,,],` and you're looking at
+    // about 48 bytes per datum. With standard downsampling of
+    // 1000 points per time series, expect ~50 KB of response payload
+    // per requested time series.
+    //
+    // Requesting 64 time series warrants a ~3 MB response, which seems
+    // reasonable.
+    const BATCH_SIZE = 64;
+
+    const requestGroups = [];
+    for (const [tag, runs] of runsByTag) {
+      for (let i = 0; i < runs.length; i += BATCH_SIZE) {
+        requestGroups.push({tag, runs: runs.slice(i, i + BATCH_SIZE)});
+      }
+    }
+
     Promise.all(
-      items.map((item) => {
-        const url = addParams(baseUrl, {tag: item.tag, run: item.run});
-        return this.requestManager
-          .request(url)
-          .then((data) => void onLoad({item, data}));
+      requestGroups.map(({tag, runs}) => {
+        return this.requestManager.request(url, {tag, runs}).then((allData) => {
+          for (const run of runs) {
+            const item = {tag, run};
+            if (Object.prototype.hasOwnProperty.call(allData, run)) {
```

Comment:
Any reason why we have to use Object.prototype.hasOwnProperty? (i.e., why not `allData.hasOwnProperty`?)

Reply:
If you have a run called `hasOwnProperty`, then that would be a type
error.

```javascript
> ({train: 1}).hasOwnProperty("train")
true
> ({train: 1, hasOwnProperty: 2}).hasOwnProperty("train")
Thrown:
TypeError: {(intermediate value)(intermediate value)}.hasOwnProperty is not a function
> Object.prototype.hasOwnProperty.call({train: 1, hasOwnProperty: 2}, "train")
true
```

Yeah.

I prefer to write this kind of code defensively—even if it seems
unlikely, it sometimes happens (e.g.: #1283).

Reply:
One day <https://github.com/hasOwnProperty> will show up in a JSON
object mapping GitHub login names to database IDs, and someone will have
a fun time…

Reply:
Got it. Thanks.